src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,131 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+import { Dispatch, SetStateAction, useRef, useState } from 'react';
+
+type Props = {
+  todo: Todo;
+  isLoading?: boolean;
+  isInEditMode?: boolean;
+  handleRemoveTodo: (todoId: number) => Promise<void>;
+  handleUpdateTodo: (todo: Todo) => Promise<void>;
+  setEditedTodoId: Dispatch<SetStateAction<null | number>>;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  isLoading,
+  handleRemoveTodo,
+  handleUpdateTodo,
+  isInEditMode,
+  setEditedTodoId,
+}) => {
+  const [todoTitleValue, setTodoTitleValue] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const onCheckTodo = () => {
+    const todoToUpdate = { ...todo, completed: !todo.completed };
+
+    handleUpdateTodo(todoToUpdate);
+  };
+
+  const onDoubleClick = () => {
+    setEditedTodoId(todo.id);
+  };
+
+  const onBlur = async (
+    event: React.FormEvent<HTMLElement> | React.FocusEvent<HTMLFormElement>,
+  ) => {
+    event.preventDefault();
+
+    const normalizedTitle = todoTitleValue.trim();
+
+    if (todo.title === normalizedTitle) {
+      setEditedTodoId(null);
+
+      return;
+    }
+
+    if (normalizedTitle === '') {
+      try {
+        await handleRemoveTodo(todo.id);
+        setEditedTodoId(null);
+      } catch (err) {}
+
+      inputRef?.current?.focus();
+
+      return;
+    }
+
+    try {
+      await handleUpdateTodo({ ...todo, title: normalizedTitle });
+      setEditedTodoId(null);
+    } catch (err) {
+      inputRef?.current?.focus();
+    }
+  };
+
+  const onKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setEditedTodoId(null);
+      setTodoTitleValue(todo.title);
+    }
+  };
+
+  return (
+    <div data-cy="Todo" className={cn('todo', { completed: todo.completed })}>
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={onCheckTodo}
+        />
+      </label>
+      {isInEditMode ? (
+        <form onSubmit={onBlur} onBlur={onBlur}>
+          <input
+            autoFocus
+            type="text"
+            data-cy="TodoTitleField"
+            className="todo__title-field"
+            placeholder="Empty todo will be deleted"
+            value={todoTitleValue}
+            onBlur={onBlur}
+            onChange={event => setTodoTitleValue(event.target.value)}
+            onKeyUp={onKeyUp}
+            ref={inputRef}
+          />
+        </form>
+      ) : (
+        <>
+          <span
+            data-cy="TodoTitle"
+            className="todo__title"
+            onDoubleClick={onDoubleClick}
+          >
+            {todo.title}
+          </span>
+          <button
+            type="button"
```

Comment:
don't forget about indents
```suggestion
      </label>
      
      {isInEditMode ? (
        <form onSubmit={onBlur} onBlur={onBlur}>
          <input
            autoFocus
            type="text"
            data-cy="TodoTitleField"
            className="todo__title-field"
            placeholder="Empty todo will be deleted"
            value={todoTitleValue}
            onBlur={onBlur}
            onChange={event => setTodoTitleValue(event.target.value)}
            onKeyUp={onKeyUp}
            ref={inputRef}
          />
        </form>
      ) : (
        <>
          <span
            data-cy="TodoTitle"
            className="todo__title"
            onDoubleClick={onDoubleClick}
          >
            {todo.title}
          </span>
          
          <button
            type="button"
```